### PR TITLE
GPII-3756: display_cluster_state

### PR DESCRIPTION
**Pull request by mrtyler**
There's a lot that can be tweaked here (running more/fewer/different commands? do we want to wait 30s * N commands if the cluster is completely broken? `-o json`?) but here is a first pass.

The `-n locust` commands are probably not very useful until I reorganize destruction (part of my plan for GPII-3756).

Sample output from stg (because I don't have a dev cluster up ATM):
```
[tyler@toaster:~/rtf/gpii-infra/gcp/live/stg]$ rake display_cluster_state
docker-compose run --rm xk rake display_cluster_state
gcloud config set project $TF_VAR_project_id
Updated property [core/project].

**************
Cluster state:
**************

timeout -t 30 kubectl -n gpii get all -o wide
NAME                                                READY     STATUS      RESTARTS   AGE       IP           NODE                                         NOMINATED NODE
pod/couchdb-couchdb-0                               2/2       Running     0          10d       10.17.1.9    gke-k8s-cluster-default-pool-bfdfbaed-g440   <none>
pod/couchdb-couchdb-1                               2/2       Running     0          10d       10.17.0.9    gke-k8s-cluster-default-pool-0dee4216-rxhd   <none>
pod/couchdb-couchdb-2                               2/2       Running     0          10d       10.17.2.16   gke-k8s-cluster-default-pool-fc3f21f1-449r   <none>
pod/couchdb-prometheus-exporter-65d7cc9cf6-wc5mn    2/2       Running     0          10d       10.17.1.11   gke-k8s-cluster-default-pool-bfdfbaed-g440   <none>
pod/dataloader-2fg48                                0/1       Completed   0          4d        10.17.0.52   gke-k8s-cluster-default-pool-0dee4216-rxhd   <none>
pod/flowmanager-76dfd4669d-cc8vg                    1/1       Running     0          4d        10.17.2.44   gke-k8s-cluster-default-pool-fc3f21f1-449r   <none>
pod/flowmanager-76dfd4669d-hzpsp                    1/1       Running     0          4d        10.17.0.54   gke-k8s-cluster-default-pool-0dee4216-rxhd   <none>
pod/flowmanager-76dfd4669d-xnpcm                    1/1       Running     0          4d        10.17.1.57   gke-k8s-cluster-default-pool-bfdfbaed-g440   <none>
pod/nginx-ingress-controller-5776579446-4mqsm       1/1       Running     0          10d       10.17.1.7    gke-k8s-cluster-default-pool-bfdfbaed-g440   <none>
pod/nginx-ingress-controller-5776579446-bgbcz       1/1       Running     0          10d       10.17.2.19   gke-k8s-cluster-default-pool-fc3f21f1-449r   <none>
pod/nginx-ingress-controller-5776579446-kp4p7       1/1       Running     1          10d       10.17.0.21   gke-k8s-cluster-default-pool-0dee4216-rxhd   <none>
pod/nginx-ingress-controller-5776579446-tcbdk       1/1       Running     2          5d        10.17.1.55   gke-k8s-cluster-default-pool-bfdfbaed-g440   <none>
pod/nginx-ingress-default-backend-597f469c8-jjvgd   1/1       Running     0          10d       10.17.0.6    gke-k8s-cluster-default-pool-0dee4216-rxhd   <none>
pod/nginx-ingress-default-backend-597f469c8-q26jd   1/1       Running     0          10d       10.17.1.8    gke-k8s-cluster-default-pool-bfdfbaed-g440   <none>
pod/preferences-7b7866b459-bcpw7                    1/1       Running     0          4d        10.17.2.45   gke-k8s-cluster-default-pool-fc3f21f1-449r   <none>
pod/preferences-7b7866b459-qsl5h                    1/1       Running     0          4d        10.17.0.53   gke-k8s-cluster-default-pool-0dee4216-rxhd   <none>
pod/preferences-7b7866b459-qtp7x                    1/1       Running     0          4d        10.17.1.56   gke-k8s-cluster-default-pool-bfdfbaed-g440   <none>

NAME                                    TYPE           CLUSTER-IP      EXTERNAL-IP     PORT(S)                      AGE       SELECTOR
service/couchdb-couchdb                 ClusterIP      None            <none>          5984/TCP                     10d       app=couchdb,release=couchdb
service/couchdb-svc-couchdb             ClusterIP      10.18.75.107    <none>          5984/TCP                     10d       app=couchdb,release=couchdb
service/flowmanager                     ClusterIP      10.18.155.126   <none>          80/TCP                       10d       app=flowmanager
service/nginx-ingress-controller        LoadBalancer   10.18.191.109   104.154.45.98   80:32262/TCP,443:30075/TCP   10d       app=nginx-ingress,component=controller,release=nginx-ingress
service/nginx-ingress-default-backend   ClusterIP      10.18.70.81     <none>          80/TCP                       10d       app=nginx-ingress,component=default-backend,release=nginx-ingress
service/preferences                     ClusterIP      10.18.196.51    <none>          80/TCP                       10d       app=preferences

NAME                                            DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE       CONTAINERS                                     IMAGES                                                                                                                                                          SELECTOR
deployment.apps/couchdb-prometheus-exporter     1         1         1            1           10d       couchdb-prometheus-exporter,prometheus-to-sd   gesellix/couchdb-prometheus-exporter@sha256:77a019a7707f581f70239783d0b76500ba25b9382d9ee0702452b0381d5722c2,gcr.io/google-containers/prometheus-to-sd:v0.3.2   app=couchdb-prometheus-exporter
deployment.apps/flowmanager                     3         3         3            3           10d       flowmanager                                    gpii/universal@sha256:1fd9d347f40d26b71e3e37132a4e4293576fde8798bcb4eaba2f291c0b899c14                                                                          app=flowmanager
deployment.apps/nginx-ingress-controller        4         4         4            4           10d       nginx-ingress-controller                       quay.io/kubernetes-ingress-controller/nginx-ingress-controller:0.15.0                                                                                           app=nginx-ingress,component=controller,release=nginx-ingress
deployment.apps/nginx-ingress-default-backend   2         2         2            2           10d       nginx-ingress-default-backend                  k8s.gcr.io/defaultbackend:1.3                                                                                                                                   app=nginx-ingress,component=default-backend,release=nginx-ingress
deployment.apps/preferences                     3         3         3            3           10d       preferences                                    gpii/universal@sha256:1fd9d347f40d26b71e3e37132a4e4293576fde8798bcb4eaba2f291c0b899c14                                                                          app=preferences

NAME                                                      DESIRED   CURRENT   READY     AGE       CONTAINERS                                     IMAGES                                                                                                                                                          SELECTOR
replicaset.apps/couchdb-prometheus-exporter-65d7cc9cf6    1         1         1         10d       couchdb-prometheus-exporter,prometheus-to-sd   gesellix/couchdb-prometheus-exporter@sha256:77a019a7707f581f70239783d0b76500ba25b9382d9ee0702452b0381d5722c2,gcr.io/google-containers/prometheus-to-sd:v0.3.2   app=couchdb-prometheus-exporter,pod-template-hash=2183775792
replicaset.apps/flowmanager-57ffb6b6ff                    0         0         0         6d        flowmanager                                    gpii/universal@sha256:cc031f92a0d02b39460f541b7e94376f2fd996ffb17c99fd1716c5d47fa46eb2                                                                          app=flowmanager,pod-template-hash=1399626299
replicaset.apps/flowmanager-6969dd7d67                    0         0         0         5d        flowmanager                                    gpii/universal@sha256:8a88f6dac65f4039d63e38036396bc6fac7fa6b7b6047096b70d3ad60e715e5a                                                                          app=flowmanager,pod-template-hash=2525883823
replicaset.apps/flowmanager-76dfd4669d                    3         3         3         4d        flowmanager                                    gpii/universal@sha256:1fd9d347f40d26b71e3e37132a4e4293576fde8798bcb4eaba2f291c0b899c14                                                                          app=flowmanager,pod-template-hash=3289802258
replicaset.apps/nginx-ingress-controller-5776579446       4         4         4         10d       nginx-ingress-controller                       quay.io/kubernetes-ingress-controller/nginx-ingress-controller:0.15.0                                                                                           app=nginx-ingress,component=controller,pod-template-hash=1332135002,release=nginx-ingress
replicaset.apps/nginx-ingress-default-backend-597f469c8   2         2         2         10d       nginx-ingress-default-backend                  k8s.gcr.io/defaultbackend:1.3                                                                                                                                   app=nginx-ingress,component=default-backend,pod-template-hash=153902574,release=nginx-ingress
replicaset.apps/preferences-5664c69b88                    0         0         0         5d        preferences                                    gpii/universal@sha256:8a88f6dac65f4039d63e38036396bc6fac7fa6b7b6047096b70d3ad60e715e5a                                                                          app=preferences,pod-template-hash=1220725644
replicaset.apps/preferences-7b7866b459                    3         3         3         4d        preferences                                    gpii/universal@sha256:1fd9d347f40d26b71e3e37132a4e4293576fde8798bcb4eaba2f291c0b899c14                                                                          app=preferences,pod-template-hash=3634226015
replicaset.apps/preferences-c96ff8b96                     0         0         0         6d        preferences                                    gpii/universal@sha256:cc031f92a0d02b39460f541b7e94376f2fd996ffb17c99fd1716c5d47fa46eb2                                                                          app=preferences,pod-template-hash=752994652

NAME                               DESIRED   CURRENT   AGE       CONTAINERS                              IMAGES
statefulset.apps/couchdb-couchdb   3         3         10d       couchdb,couchdb-statefulset-assembler   couchdb:2.2.0,kocolosk/couchdb-statefulset-assembler:1.1.0

NAME                                                           REFERENCE                             TARGETS           MINPODS   MAXPODS   REPLICAS   AGE
horizontalpodautoscaler.autoscaling/nginx-ingress-controller   Deployment/nginx-ingress-controller   47%/50%, 3%/50%   2         4         4          10d

NAME                   DESIRED   SUCCESSFUL   AGE       CONTAINERS        IMAGES                                                                                   SELECTOR
job.batch/dataloader   1         1            4d        gpii-dataloader   gpii/universal@sha256:1fd9d347f40d26b71e3e37132a4e4293576fde8798bcb4eaba2f291c0b899c14   controller-uid=b6c22adb-3619-11e9-a03f-42010a8000db
timeout -t 30 kubectl -n gpii get pv -o wide
NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS    CLAIM                                     STORAGECLASS   REASON    AGE
pvc-9f68ef52-fd37-11e8-8c39-42010a80009f   10Gi       RWO            Delete           Bound     gpii/database-storage-couchdb-couchdb-0   pd-ssd                   10d
pvc-9f829990-fd37-11e8-8c39-42010a80009f   10Gi       RWO            Delete           Bound     gpii/database-storage-couchdb-couchdb-1   pd-ssd                   10d
pvc-9f91dcfe-fd37-11e8-8c39-42010a80009f   10Gi       RWO            Delete           Bound     gpii/database-storage-couchdb-couchdb-2   pd-ssd                   10d
timeout -t 30 kubectl -n gpii get pvc -o wide
NAME                                 STATUS    VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
database-storage-couchdb-couchdb-0   Bound     pvc-9f68ef52-fd37-11e8-8c39-42010a80009f   10Gi       RWO            pd-ssd         10d
database-storage-couchdb-couchdb-1   Bound     pvc-9f829990-fd37-11e8-8c39-42010a80009f   10Gi       RWO            pd-ssd         10d
database-storage-couchdb-couchdb-2   Bound     pvc-9f91dcfe-fd37-11e8-8c39-42010a80009f   10Gi       RWO            pd-ssd         10d
timeout -t 30 kubectl -n gpii get events -o wide
LAST SEEN   FIRST SEEN   COUNT     NAME                                 KIND      SUBOBJECT                  TYPE      REASON      SOURCE                                                MESSAGE
54m         54m          1         couchdb-couchdb-1.1586f76ce476d370   Pod       spec.containers{couchdb}   Warning   Unhealthy   kubelet, gke-k8s-cluster-default-pool-0dee4216-rxhd   Liveness probe failed: rpc error: code = 2 desc = oci runtime error: exec failed: container_linux.go:247: starting container process caused "process_linux.go:87: adding pid 3684298 to cgroups caused \"failed to write 3684298 to cgroup.procs: write /sys/fs/cgroup/cpu,cpuacct/kubepods/burstable/pod2ecbdd51-3173-11e9-bb70-42010a800085/fa95a5e6684341606d9da022677de872e4f30f196c68e90b9814fab5189a5702/cgroup.procs: invalid argument\""


5m        4d        947       flowmanager-ingress.1585bc775a0e52b1   Ingress             Warning   Translate   loadbalancer-controller   error while evaluating the ingress spec: service "gpii/flowmanager" is type "ClusterIP", expected "NodePort" or "LoadBalancer"
5m        4d        948       preferences-ingress.1585bc775a0cf3d6   Ingress             Warning   Translate   loadbalancer-controller   error while evaluating the ingress spec: service "gpii/preferences" is type "ClusterIP", expected "NodePort" or "LoadBalancer"
```

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -57,7 +57,7 @@ gcp-setup:
   script:
     - docker version
     - docker-compose version
-    - docker pull gpii/exekube:0.4.0-google
```
Comment (stepanstipl) [inline .gitlab-ci.yml]: Is this temporary till we get proper versioning for exekube?

Comment (mrtyler) [reply .gitlab-ci.yml]: This is updating gcp-setup from an older approach to the current state of the art. Updating Exekube versioning is in https://issues.gpii.net/browse/GPII-3745.

**Comment by mrtyler** [comment]
> I typically find the `json` output more useful, as it contains much more info.

Indeed. This is not entirely scientific but for a sense of scale:

```
[tyler@toaster:~/rtf/gpii-infra/gcp/live/stg]$ rake display_cluster_state | wc  # no -o flag
      80     460    7347

[tyler@toaster:~/rtf/gpii-infra/gcp/live/stg]$ rake display_cluster_state | wc  # -o wide
      80     592   13538

[tyler@toaster:~/rtf/gpii-infra/gcp/live/stg]$ rake display_cluster_state | wc  # -o json
    6871   11749  313642
```

> what about saving output (long json form, from all namespaces) somewhere? Does Gitlab have any feature like that?

I think [Artifacts](https://docs.gitlab.com/ee/user/project/pipelines/job_artifacts.html) are the relevant feature. It's a little tricky because we can't write files inside the container without using a Docker volume (see https://github.com/gpii-ops/gpii-infra/pull/92 as a start). One workaround is to run the commands inside the exekube container, but redirect the output outside the container (i.e. in entrypoint.rake directly).

> Perhaps this can be in a form of rake task that would upload the logs somewhere (run always from CI, possible to run by user in trouble on request)?

Sure, maybe to a GCS bucket, maybe the exported-logs bucket we already have in each Project.

These are good suggestions and I'd like to continue the conversation (in this PR or elsewhere). However, I'm going to merge this now so we can start gathering data and to make display_cluster_state available (and to make the branches less confusing for some upcoming destroy changes).

**Comment by natarajaya** [comment]
LGTM

**Comment by stepanstipl** [comment]
> Sure, maybe to a GCS bucket, maybe the exported-logs bucket we already have in each Project.

I think I would prefer GCS bucket as well (also to avoid hassle with Artifacts feature), and then this would be consistent both for CI runs and user-executed report.

One question here - aren't the buckets deleted atm on infra teardown? Maybe some shared bucket for this purpose?

**Comment by mrtyler** [comment]
> aren't the buckets deleted atm. on infra teardown?

Ah, yes they are.

> Maybe some shared bucket for this purpose outside of the individual projects?

Maybe. There would be some details to work out, e.g. how would permissions work (one bucket per environment?)?

**Comment by stepanstipl** [comment]
There's pretty good command: `kubectl   cluster-info dump` that gives you a lot of stuff.

> how would permissions work

We can create a "common" bucket and set iam to allow only uploading new objects from anyone within rtf org (`storage.objects.create` permission only) and just use `gsutil` to copy files to some structure like `/<env>/<user>/<timestamp>.dump` or smth.

I think that way we would not expose anything and only admins would have permissions to read the content. Buckets are encrypted, so even if there would be some potentially sensitive stuff it should be ok.

And I would create lifecycle mgmt policy on the bucket to auto-delete objects after let's say 30 days.

**Pull request merged**